### ISSUE TITLE
[vue-template-compiler] Upgrading VueJS from 2.0.3 to 2.0.4 freezes build process (using vue-loader)

### ISSUE THREAD

**Issue by jbruni** [vuejs/vue#4127]
### Vue.js version
2.0.4

### Reproduction Link
It is happening in a big and complex project.
I will see if I can shrink down to a reproduceable scenario.

### Steps to reproduce
- Upgrade from Vue 2.0.3 to 2.0.4
- Run build process

### What is Expected?
- Build process to finish normally

### What is actually happening?
- It freezes (interrupting with CTRL+C returns to CLI prompt)

-----------------

My scenario is complex, it uses Gulp, Webpack, Babel, vue-loader, etc. I updated `package.json` from VueJS 2.0.3 to 2.0.4 and my build task stucks. If I revert, it runs fine.

I trimmed the issue down to `vue-template-compiler`.

If I manually **switch back vue-template-compiler folder to version 2.0.3** inside node_modules, and run with everything else exactly the same (VueJS 2.0.4), nothing else changed, it gives me that warning error message telling me to "bump to latest"; but if I go further and force it to run with mismatching versions (by changing [this line from vue-template-compiler](https://github.com/vuejs/vue/blob/ae93d2bef8e20402ae3501aa5cdcf14b0ee19c12/packages/vue-template-compiler/index.js#L7) to `if (false) {`), then the build completes.

In short: there is an issue at `vue-template-compiler` 2.0.4 freezing the compilation of some template in my project.

Note: I didn't try to run the compiled result. I believe result is unpredictable. The fact is: same codebase compiles fine with vue-template-compiler 2.0.3, but the build process freezes with vue-template-compiler 2.0.4

I will try to get more info and eventually something to reproduce.

Thank you.

**Comment by marcim** [comment]
Same here!

Webpack freezes when I run `gulp`.
I'm using Yarn, I just tried to reinstall all dependencies, but without success.
Back to 2.0.3 worked!

Thanks!

**Comment by yyx990803** [comment]
@jbruni can you try to build Vue from latest source and see if that resolves the issue?

**Comment by yyx990803** [comment]
Ok, I think I know where the problem is, will fix soon

**Comment by yyx990803** [comment]
@jbruni pushed fix, please try to build from dev branch source and let me know

**Comment by jbruni** [comment]
Thanks, Evan. I'm at home now. I'll be back in the office tomorrow. I'll test; first thing in the morning.

**Comment by jbruni** [comment]
@yyx990803 - Just tested using the 2.0.5 release and it works. Thanks a lot for your amazing work.

I sincerely got curious to know which *edge case* is the case of any of my templates which got caught by the issue...
